subsys/net/lib/tls_credentials/tls_credentials_shell.c

```diff
@@ -0,0 +1,805 @@
+						 "be leaked if this credential is deleted without "
+						 "using the shell.\n", cred_written);
+
+		cred_chosen = cred_copy;
+	}
+
+	/* Finally, store the credential in whatever credentials backend is active. */
+	err = tls_credential_add(sectag, type, cred_chosen, cred_written);
+	if (err) {
+		shell_fprintf(sh, SHELL_ERROR, "Failed to add TLS credential with sectag %d and "
+					       "type %s. Error: %d\n.", sectag,
+					       cred_type_name(type), err);
+		goto cleanup;
+	}
+
+	/* Do not free the copied key during cleanup, since it was successfully written. */
+	keep_copy = true;
+
+	shell_fprintf(sh, SHELL_NORMAL, "Added TLS credential of type %s, sectag %d, and length %d "
+					"bytes.\n", cred_type_name(type), sectag, cred_written);
+
+cleanup:
+	/* Unlock credentials since we are done interacting with internal state. */
+	credentials_unlock();
+
+	/* We are also done with the credentials buffer, so clear it for good measure. */
+	shell_clear_cred_buf(sh);
+
+	/* If we copied the credential, make sure it is eventually freed. */
+	if (cred_copy) {
+		if (keep_copy) {
+			/* If the credential was successfully stored, keep a reference to it in case
+			 * it is ever deleted and needs to be freed.
+			 */
+			cred_refs[ref_slot] = cred_copy;
+		} else {
+			/* Otherwise, clear and free it immediately */
+			memset(cred_copy, 0, cred_written);
+			(void)k_free(cred_copy);
+		}
+	}
+
+	return err;
+}
+
+/* Buffers credential data into the credential buffer. */
+static int tls_cred_cmd_buf(const struct shell *sh, size_t argc, char *argv[])
+{
+	/* If the "clear" keyword is provided, clear the buffer rather than write to it. */
+	if (strcmp(argv[1], "clear") == 0) {
+		shell_clear_cred_buf(sh);
+		return 0;
+	}
+
+	/* Otherwise, assume provided arg is base64 and attempt to write it into the credential
+	 * buffer.
+	 */
+	return shell_write_cred_buf(sh, argv[1]);
+}
+
+/* Deletes a credential from the credential store */
+static int tls_cred_cmd_del(const struct shell *sh, size_t argc, char *argv[])
+{
+	int err = 0;
+	sec_tag_t sectag;
+	enum tls_credential_type type;
+	struct tls_credential *cred = NULL;
+	int ref_slot = -1;
+
+	/* Lock credentials so that we can safely use internal access functions. */
+	credentials_lock();
+
+	err = shell_parse_cred_sectag(sh, argv[1], &sectag, false);
+	if (err) {
+		goto cleanup;
+	}
+
+	err = shell_parse_cred_type(sh, argv[2], &type, false);
+	if (err) {
+		goto cleanup;
+	}
+
+	/* Check whether a credential of this type and sectag actually exists. */
+	cred = credential_get(sectag, type);
+	if (!cred) {
+		shell_fprintf(sh, SHELL_ERROR, "There is no TLS credential with sectag %d and "
+					       "type %s.\n", sectag, cred_type_name(type));
+		err = -ENOENT;
+		goto cleanup;
+	}
+
+	ref_slot = find_ref_slot(cred->buf);
+	if (ref_slot >= 0) {
+		/* This was a credential we copied to heap. Clear and free it. */
```

Comment:
Shouldn't you clear the cred_ref once you free the memory, so it is an empty "ref slot" again that can be reused?

Reply:
gah, you are absolutely correct, I forgot to do this. 🤦